### ISSUE TITLE
Runtime recompilation fails with CS8858 if with statement used on a non-local record type.

### ISSUE THREAD

**Issue by markdchurchill** [dotnet/aspnetcore#27756]
### Describe
Given a record type declared outside of a razor page, runtime recompilation fails if a `with` statement is used inside the page.

### To Reproduce
Take a Razor Pages project with ...Razor.RuntimeCompilation enabled.

Add a simple record type to the assembly: `public record Foo(int Bar);`

Use the 'with' statement: `var foo = new Foo(1) with { Bar = 2 };`

Run the app, edit the page, refresh.

### Exceptions (if any)
```
 error CS8858: The receiver type 'Foo' is not a valid record type.
         at Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.RuntimeViewCompiler.CompileAndEmit(
RazorCodeDocument codeDocument, String generatedCode)
```

### Extra

There is no issue if you declare the record type inline in a `@function` block.

### Further technical details
aspnet packages: 5.0.0

Rider IDE

.NET SDK (reflecting any global.json):
 Version:   5.0.100
 Commit:    5044b93829

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100\

Host (useful for support):
  Version: 5.0.0
  Commit:  cf258a14b7

**Comment by mustakimali** [comment]
This also happens if you are trying to use a `record` type that is defined in a Nuget package. Happening right after updating a project from `netcoreapp3.1` to `net5`.

**Comment by mkArtakMSFT** [comment]
@ajaybhargavb can you please investigate this? Need to understand whether this behaves correctly in a console app. If not - then this needs to move to the compiler team.

**Comment by ajaybhargavb** [comment]
Just investigated this and this indeed seems to be a bug. I am not able to reproduce the same thing in a regular console app. We're possibly not including all compilation references during recompilation. cc @pranavkm who has more context.

**Comment by mkArtakMSFT** [comment]
Closing as the runtime compilation option is not available in 6.0 any more.
